src/openrct2-ui/ride/Construction.cpp

```diff
@@ -36,3 +40,60 @@ void RideConstructNew(RideSelection listItem)
 
     GameActions::Execute(&gameAction);
 }
+
+SpecialElementsDropdownState BuildSpecialElementsList(
+    const Ride& currentRide, uint8_t buildDirection, uint8_t buildSlope, uint8_t buildBank, RideConstructionState state)
+{
+    auto buildDirectionIsDiagonal = TrackPieceDirectionIsDiagonal(buildDirection);
+    SpecialElementsDropdownState list;
+
+    // if it's building neither forwards nor backwards, no list is available
+    if (state != RideConstructionState::Front && state != RideConstructionState::Place && state != RideConstructionState::Back)
+        return list;
+
+    for (track_type_t trackType : DropdownOrder)
+    {
+        const auto& ted = GetTrackElementDescriptor(trackType);
+        if (!IsTrackEnabled(ted.Definition.type))
+            continue;
+        bool entryIsDisabled;
+
+        // If the current build orientation (slope, bank, diagonal) matches the track element's, show the piece as enabled
+        if (state == RideConstructionState::Back)
+        {
+            entryIsDisabled = ted.Definition.vangle_end != buildSlope || ted.Definition.bank_end != buildBank
+                || TrackPieceDirectionIsDiagonal(ted.Coordinates.rotation_end) != buildDirectionIsDiagonal;
+        }
+        else
+        {
+            entryIsDisabled = ted.Definition.vangle_start != buildSlope || ted.Definition.bank_start != buildBank
+                || TrackPieceDirectionIsDiagonal(ted.Coordinates.rotation_begin) != buildDirectionIsDiagonal;
+        }
+
+        // Additional tower bases can only be built if the ride allows for it (elevator)
+        if (trackType == TrackElemType::TowerBase
+            && !currentRide.GetRideTypeDescriptor().HasFlag(RIDE_TYPE_FLAG_ALLOW_EXTRA_TOWER_BASES))
+            entryIsDisabled = true;
+
+        // Check if a previous element exists, to collate entries if possible
+        if (list.Elements.size() > 0
+            && GetTrackElementDescriptor(list.Elements.back().TrackType).Description == ted.Description)
+        {
+            // If the current element is disabled, do not add current element.
+            if (entryIsDisabled)
+                continue;
+            // If the previous element is disabled and current element is enabled, replace the previous element
+            if (list.Elements.back().Disabled && !entryIsDisabled)
```

Comment:
```suggestion
            auto& lastElement = list.Elements.back();
            if (lastElement.Disabled && !entryIsDisabled)
```
the lines below should use the variable as well.